sci-libs/vtk: add 9.2.2

waebbl opened the pull request:
- add a new function to check memory and disk space requirements
- don't drop bundled external libraries, it's only a few MB which are freed
	through this and upstream does a good job to avoid automagic deps
- request more modules to build explicitly depending on USE flag settings
- don't export CC and friends with USE=mpi, package builds fine with gcc and still
	links to MPI libraries
- add USE flag to build vtk-m optional

Bug: https://bugs.gentoo.org/793221
Bug: https://bugs.gentoo.org/835659
Bug: https://bugs.gentoo.org/880759
Closes: https://bugs.gentoo.org/820593
Closes: https://bugs.gentoo.org/827310
Closes: https://bugs.gentoo.org/846374
Closes: https://bugs.gentoo.org/874798
Signed-off-by: Bernd Waibel <waebbl-gentoo@posteo.net>

waebbl commented:
Converting to draft, feel free to review and/or test.

juippis reviewed (commented):
```
CMake Warning at CMake/vtkModule.cmake:4572 (find_package):
  By not providing "Findospray.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "ospray", but
  CMake did not find one.

  Could not find a package configuration file provided by "ospray" (requested
  version 2.1) with any of the following names:

    osprayConfig.cmake
    ospray-config.cmake

  Add the installation prefix of "ospray" to CMAKE_PREFIX_PATH or set
  "ospray_DIR" to a directory containing one of the above files.  If "ospray"
  provides a separate development package or SDK, be sure it has been
  installed.
Call Stack (most recent call first):
  Rendering/RayTracing/CMakeLists.txt:58 (vtk_module_find_package)


CMake Error at CMake/vtkModule.cmake:4578 (message):
  Could not find the ospray external dependency.
Call Stack (most recent call first):
  Rendering/RayTracing/CMakeLists.txt:58 (vtk_module_find_package)
```
```
CMake Warning:
  Manually-specified variables were not used by the project:

    VTK_GROUP_ENABLE_Tk

```
```
ninja: error: '/usr/lib64/libzip.so', needed by 'lib64/libvtkexodusII-9.2.so.9.2.2', missing and no known rule to make it
```

waebbl commented on sci-libs/vtk/vtk-9.2.2.ebuild:
```diff
@@ -0,0 +1,742 @@
+	${RDEPEND}
+	dev-cpp/eigen
+	dev-cpp/nlohmann_json
+	<dev-libs/pegtl-3
```
Yes it's still true. I had a link to it once, will try to find it again and state it as requested.

waebbl commented:
> ```
> CMake Warning at CMake/vtkModule.cmake:4572 (find_package):
>   By not providing "Findospray.cmake" in CMAKE_MODULE_PATH this project has
> 
> ninja: error: '/usr/lib64/libzip.so', needed by 'lib64/libvtkexodusII-9.2.so.9.2.2', missing and no known rule to make it
> ```

Started yesterday to get back to this again and test the debug and test uses.

Strange enough, the first error about OSPRay is new on the default and even minimal build, as well as the GROUP_Tk message. I changed these already, because they showed on my tests yesterday as well.

Never saw the ninja error however. Was it the default USE configuration which produced it?

juippis commented:
> 
> Never saw the ninja error however. Was it the default USE configuration which produced it?
> 
> dev-libs/libzip is still in the comment section for the deps: https://github.com/gentoo/gentoo/pull/28250/files#diff-ee45605f85b867a8c226e60f01e693598a3737e0ae1cff9ce06bb210c79bacb9R72
> 

USE:        abi_x86_64 amd64 elibc_glibc freetype imaging kernel_linux kits las logging mysql postgres postgres_targets_postgres10 postgres_targets_postgres11 postgres_targets_postgres13                       postgres_targets_postgres14 python python_single_target_python3_9 tbb userland_GNU web

juippis commented on sci-libs/vtk/vtk-9.2.2.ebuild:
```diff
@@ -0,0 +1,742 @@
+		ewarn "See bug #820593"
+	fi
+
+	if use doc || use examples; then
+		CHECKREQS_DISK_BUILD="7G"
+	fi
```
I'm not sure why, but this doesn't seem to work properly:
```
 * Package:    sci-libs/vtk-9.2.2:0/9.2
 * Repository: gentoo
 * Maintainer: sci@gentoo.org waebbl-gentoo@posteo.net,proxy-maint@gentoo.org
 * USE:        abi_x86_64 amd64 doc elibc_glibc ffmpeg freetype gdal imaging java kernel_linux kits las   mpi mysql openvdb postgres postgres_targets_postgres10 postgres_targets_postgres12                        python_single_target_python3_9 rendering sdl userland_GNU
 * FEATURES:   network-sandbox preserve-libs sandbox userpriv usersandbox
/var/db/repos/gentoo/sci-libs/vtk/vtk-9.2.2.ebuild: line 223: examples: command not found
 * Checking for at least 7 GiB disk space at "/var/tmp/portage/sci-libs/vtk-9.2.2/temp" ...
 [ ok ]
 * Using: openjdk-bin-17
 * PostgreSQL Target: 12
>>> Unpacking source...
>>> Unpacking VTK-9.2.2.tar.gz to /var/tmp/portage/sci-libs/vtk-9.2.2/work
>>> Unpacking VTKData-9.2.2.tar.gz to /var/tmp/portage/sci-libs/vtk-9.2.2/work
>>> Unpacking VTKDataFiles-9.2.2.tar.gz to /var/tmp/portage/sci-libs/vtk-9.2.2/work
>>> Unpacking vtkDocHtml-9.2.2.tar.gz to /var/tmp/portage/sci-libs/vtk-9.2.2/work
>>> Source unpacked in /var/tmp/portage/sci-libs/vtk-9.2.2/work
>>> Preparing source in /var/tmp/portage/sci-libs/vtk-9.2.2/work/VTK-9.2.2 ...
```
> /var/db/repos/gentoo/sci-libs/vtk/vtk-9.2.2.ebuild: line 223: examples: command not found

waebbl commented on sci-libs/vtk/vtk-9.2.2.ebuild:
```diff
@@ -0,0 +1,742 @@
+
+	if use doc && examples; then
+			CHECKREQS_DISK_BUILD="10G"
+	fi
```
The error about examples is here. I hit it as well yesterday and changed it already.

waebbl commented:
> > Never saw the ninja error however. Was it the default USE configuration which produced it?
> > dev-libs/libzip is still in the comment section for the deps: https://github.com/gentoo/gentoo/pull/28250/files#diff-ee45605f85b867a8c226e60f01e693598a3737e0ae1cff9ce06bb210c79bacb9R72
> 
> USE: abi_x86_64 amd64 elibc_glibc freetype imaging kernel_linux kits las logging mysql postgres postgres_targets_postgres10 postgres_targets_postgres11 postgres_targets_postgres13 postgres_targets_postgres14 python python_single_target_python3_9 tbb userland_GNU web

Thanks, I check these flags and see if one specific requires libzip.

waebbl commented:
About using postgres.eclass: PkgCheck now seems to require the postgres eclass to be used. A simple depstring like `>=dev-db/postgresql-10` without using a slot operator leads to a warning from pkgcheck. Some questions / points about this:
1. postgres.eclass is EAPI 7, so the ebuild needs to downgrade to EAPI 7 as well
2. VTK does not support several postgres installations at once and will use the implementation which is set by `eselect postgresql`, thus rendering several postgres targets defined mostly meaningless. It needs >=postgresql-10, no other restrictions apply by the build system.
3. By default, postgres.eclass has two targets active, IIRC 11 and 12. Although this is a thing with the postgres eclass, it looks like it's not supporting a single targeting like we do it with Python. Or do I miss a thing which I have to call in one of the early build functions?

How to best handle these things? Is it ok the way it's currently done, i.e. depending on `${POSTGRES_DEPS}` and leave the rest to the eclasses, or do I need to implement multiprocessing aware functions to use the postgres eclass?

waebbl commented:
> > > Never saw the ninja error however. Was it the default USE configuration which produced it?
> > > dev-libs/libzip is still in the comment section for the deps: https://github.com/gentoo/gentoo/pull/28250/files#diff-ee45605f85b867a8c226e60f01e693598a3737e0ae1cff9ce06bb210c79bacb9R72
> > 
> > 
> > USE: abi_x86_64 amd64 elibc_glibc freetype imaging kernel_linux kits las logging mysql postgres postgres_targets_postgres10 postgres_targets_postgres11 postgres_targets_postgres13 postgres_targets_postgres14 python python_single_target_python3_9 tbb userland_GNU web
> 
> Thanks, I check these flags and see if one specific requires libzip.

Don't understand this issue and can't reproduce it, even with the settings you used. I just re-add libzip for now to avoid.
The error suprises me actually, because the ThirdParty/exodusII/vtk.module lists only netcdf and hdf5 as dependencies. The netcdf lists hdf5 and hdf5 lists zlib as dependencies. There should be no need for libzip in the libvtkexodus library.

juippis commented:
> About using postgres.eclass: PkgCheck now seems to require the postgres eclass to be used. A simple depstring like `>=dev-db/postgresql-10` without using a slot operator leads to a warning from pkgcheck. Some questions / points about this:
> 
>     1. postgres.eclass is EAPI 7, so the ebuild needs to downgrade to EAPI 7 as well
> 
>     2. VTK does not support several postgres installations at once and will use the implementation which is set by `eselect postgresql`, thus rendering several postgres targets defined mostly meaningless. It needs >=postgresql-10, no other restrictions apply by the build system.
> 
>     3. By default, postgres.eclass has two targets active, IIRC 11 and 12. Although this is a thing with the postgres eclass, it looks like it's not supporting a single targeting like we do it with Python. Or do I miss a thing which I have to call in one of the early build functions?
> 
> 
> How to best handle these things? Is it ok the way it's currently done, i.e. depending on `${POSTGRES_DEPS}` and leave the rest to the eclasses, or do I need to implement multiprocessing aware functions to use the postgres eclass?

What's the pkgcheck warning? Nothing comes to mind why *postgres.eclass* or postgres dep should be treated specially. I imagine you can just do: `>=dev-db/postgresql-10:*` and strip out all of the eclass stuff. (Note: :* will only work if nothing links to postresql, use `:=` if something does)

juippis commented:
> > > > Never saw the ninja error however. Was it the default USE configuration which produced it?
> > > > dev-libs/libzip is still in the comment section for the deps: https://github.com/gentoo/gentoo/pull/28250/files#diff-ee45605f85b867a8c226e60f01e693598a3737e0ae1cff9ce06bb210c79bacb9R72
> > > 
> > > 
> > > USE: abi_x86_64 amd64 elibc_glibc freetype imaging kernel_linux kits las logging mysql postgres postgres_targets_postgres10 postgres_targets_postgres11 postgres_targets_postgres13 postgres_targets_postgres14 python python_single_target_python3_9 tbb userland_GNU web
> > 
> > 
> > Thanks, I check these flags and see if one specific requires libzip.
> 
> Don't understand this issue and can't reproduce it, even with the settings you used. I just re-add libzip for now to avoid. The error suprises me actually, because the ThirdParty/exodusII/vtk.module lists only netcdf and hdf5 as dependencies. The netcdf lists hdf5 and hdf5 lists zlib as dependencies. There should be no need for libzip in the libvtkexodus library.

It starting to sound like it's some automagic linking issue somewhere... *sigh*

waebbl commented:
> 
> What's the pkgcheck warning? Nothing comes to mind why *postgres.eclass* or postgres dep should be treated specially. I imagine you can just do: `>=dev-db/postgresql-10:*` and strip out all of the eclass stuff. (Note: :* will only work if nothing links to postresql, use `:=` if something does)

Im on vacation this Week. Don't remember the exact error, it was something about no slot being used. I check on weekend, once I'm back at home.

waebbl commented:
> What's the pkgcheck warning? Nothing comes to mind why _postgres.eclass_ or postgres dep should be treated specially. I imagine you can just do: `>=dev-db/postgresql-10:*` and strip out all of the eclass stuff. (Note: :* will only work if nothing links to postresql, use `:=` if something does)

Sadly I couldn't reproduce the warning, so I went back to the depstring I used in 9.1 without having this issue anymore and removed the postgres.eclass related stuff.

Package should be ready. Some tests are failing, it's a minor number compared to the amount of tests shipped. On a full build, i.e. maximum use flag setting, ~70 out of ~2.5k tests fail. I have not yet figured out what's the reason for this, because on their dashboard usually all tests succeed in a virtual environment.

juippis commented:
[vtk-9.2.2:20221205-114840.log](https://github.com/gentoo/gentoo/files/10154871/vtk-9.2.2.20221205-114840.log)

With these use flags something weird happened. 5/6 runs were successful, tests failed as you mentioned. Also, this package was absolutely killing my pc at one point:
```
>>> Emerging (19 of 19) sci-libs/vtk-9.2.2::gentoo
>>> Jobs: 18 of 19 complete, 1 running              Load avg: 227, 155, 96
```
o.o

waebbl commented:
> [vtk-9.2.2:20221205-114840.log](https://github.com/gentoo/gentoo/files/10154871/vtk-9.2.2.20221205-114840.log)
> 
> With these use flags something weird happened. 5/6 runs were successful, tests failed as you mentioned. Also, this package was absolutely killing my pc at one point:
> 
> ```
> >>> Emerging (19 of 19) sci-libs/vtk-9.2.2::gentoo
> >>> Jobs: 18 of 19 complete, 1 running              Load avg: 227, 155, 96
> ```
> 
> o.o

This looks weird. Maybe I'm passing too many arguments to cmake which conflict for some use flag settings. I added several options being passed to the default and rendering uses especially. I'm gonna test this.

The load is probably related to USE=vtkm, it can grow more if the cuda use flag is added. This and the long build times for vtkm, especially when used with cuda were the main reason, why I added a use flag for this module. Although I haven't considered such high load on my machine.

juippis commented:
I believe I noticed high loads with:

> abi_x86_64 amd64 doc elibc_glibc gdal imaging java kernel_linux las logging mpi mysql odbc openvdb pdal postgres python python_single_target_python3_8 rendering sdl tbb threads userland_GNU

Could have happened later too, I left my pc alone at some point. I saw it turned on java, mpi and gdal. imaging was also a bit suspectible.

waebbl commented:
> Could have happened later too, I left my pc alone at some point. I saw it turned on java, mpi and gdal. imaging was also a bit suspectible.

Java could also be a candidate for it, because it doesn't respect `{MAKE,NINJA}OPTS` and always uses all available cores.

waebbl commented:
> > [vtk-9.2.2:20221205-114840.log](https://github.com/gentoo/gentoo/files/10154871/vtk-9.2.2.20221205-114840.log)
> > With these use flags something weird happened. 5/6 runs were successful, tests failed as you mentioned. Also, this package was absolutely killing my pc at one point:
> > ```
> > >>> Emerging (19 of 19) sci-libs/vtk-9.2.2::gentoo
> > >>> Jobs: 18 of 19 complete, 1 running              Load avg: 227, 155, 96
> > ```
> > o.o
> 
> This looks weird. Maybe I'm passing too many arguments to cmake which conflict for some use flag settings. I added several options being passed to the default and rendering use-cases especially. I'm gonna test this.
> 
> The load is probably related to USE=vtkm, it can grow more if the cuda use flag is added. This and the long build times for vtkm, especially when used with cuda were the main reason, why I added a use flag for this module. Although I haven't considered such high load on my machine.

@juippis This is due to the `kits` USE flag. I now remember having such a conflict in 9.1.0, which is why I removed the USE flag in that version in the first place.

Considering the minor effect of this option, I'm going to remove the USE flag from this version as well. I honestly don't have a good idea, how these inter-module-dependencies could be thoroughly analyzed and reflected in the ebuild.

juippis reviewed (approved):
Thanks!

waebbl commented:
Thanks for your patience!
